syspage.c

```diff
@@ -13,183 +13,83 @@
  * %LICENSE%
  */
 
-
 #include "syspage.h"
 
 #include <hal/hal.h>
 #include <lib/log.h>
-#include <lib/console.h>
+#include <lib/list.h>
 #include <lib/errno.h>
 
 
-#define MAX_PROGRAMS_NB   32
-#define MAX_MAPS_NB       16
-#define MAX_MAP_NAME_SIZE 8
-#define MAX_ARGS_SIZE     256
-#define MAX_ENTRIES_NB    6 /* 3 of kernel's sections, 2 of plo's sections and syspage */
-
-#define MAX_SYSPAGE_SIZE (sizeof(syspage_t) + MAX_ARGS_SIZE + MAX_PROGRAMS_NB * sizeof(syspage_program_t) + MAX_MAPS_NB * sizeof(syspage_map_t))
-
-
-#pragma pack(push, 1)
-
-typedef struct {
-	addr_t start;
-	addr_t end;
-	u32 attr;
-
-	u8 id;
-	char name[SIZE_MAP_NAME + 1];
-} syspage_map_t;
-
-
-typedef struct {
-	addr_t start;
-	addr_t end;
-
-	u8 dmap;
-	u8 imap;
-
-	char name[SIZE_APP_NAME + 1];
-} syspage_program_t;
-
-
-typedef struct {
-	syspage_hal_t hal;
-
-	struct {
-		addr_t text;
-		size_t textsz;
-
-		addr_t data;
-		size_t datasz;
-
-		addr_t bss;
-		size_t bsssz;
-	} kernel;
-
-	size_t syspagesz;
-
-	char *arg;
+#define ALIGN_ADDR(addr, align) (align ? ((addr + (align - 1)) & ~(align - 1)) : addr)
 
-	size_t progssz;
-	syspage_program_t *progs;
-
-	size_t mapssz;
-	syspage_map_t *maps;
-} syspage_t;
-
-#pragma pack(pop)
-
-
-typedef struct {
-	addr_t start;
-	addr_t end;
-} map_entry_t;
-
-
-/* Structure contains entries which belong to syspage map and occupied its space */
-typedef struct {
-	map_entry_t entry[MAX_ENTRIES_NB];
-
-	addr_t top;
-	syspage_map_t map;
-} plo_map_t;
-
-
-/* The syspage_common contains fields which are copied into to the kernel's memory.
- * The *syspage points to memory which is set in run time and is shared with plo and kernel. */
 struct {
 	syspage_t *syspage;
 
-	size_t argCnt;
-	char args[MAX_ARGS_SIZE];
-
-	size_t progsCnt;
-	syspage_program_t progs[MAX_PROGRAMS_NB];
+	void *heapTop;
+	void *heapEnd;
+} syspage_common;
 
-	size_t mapsCnt;
-	plo_map_t maps[MAX_MAPS_NB];
 
-	addr_t kernel_entry;
+/* General functions */
 
-	/* General entries: syspage, kernel's elf sections and plo's elf sections */
-	map_entry_t entries[MAX_ENTRIES_NB];
+void syspage_init(void)
+{
+	/* Syspage is placed above .bss and aligned to page size.
+	 * First part of the syspage is owned by HAL, it is described by 'syspage_hal_t' struct */
+	syspage_common.syspage = (syspage_t *)((char *)ADDRESS_SYSPAGE + sizeof(syspage_hal_t));
 
-	syspage_hal_t hal;
-} syspage_common;
+	syspage_common.syspage->maps = NULL;
+	syspage_common.syspage->progs = NULL;
+	syspage_common.syspage->size = sizeof(syspage_t);
 
+	syspage_common.heapTop = (void *)((char *)syspage_common.syspage + sizeof(syspage_t));
+	syspage_common.heapEnd = (void *)((char *)syspage_common.syspage + SIZE_SYSPAGE);
```

Comment:
Do we have to explicitly cast type to void *? I'd remove the (void *) cast in other places too.

Reply:
It's not necessary, I wanted to make it more clear. 
I've removed casting to `void *` in assignments.